### ISSUE TITLE
add a lookup table into the PMML

### ISSUE THREAD

**Issue by jibybabu** [jpmml/sklearn2pmml#14]
Hi,

I was wondering if there is any way  i can embed a lookup table into the PMML file using sklearn2pmml. For example, My model is having 40 features.
Out of 40, 10 features i am filling using a CSV file which is of shape 100000 X 10 .
I would like to attach this info to the PMML file sothat when the model is running, it will take the value from there. 

Thanks,
Jiby

**Comment by vruusmann** [comment]
A lookup table (between two categorical value spaces) is represented in PMML using the [`MapValues` element](http://dmg.org/pmml/v4-3/Transformations.html#xsdElement_MapValues). The `MapValues` transformation can work with external data sources (such as CSV files or SQL databases), but most of the time it's easier if the data is inlined. The choice between those two options depends on how frequently/on what extent this lookup table needs to be updated, and by whom.

In your case, if those ten features are independent of one another, then you should define ten independent `MapValues` transformations (each with one input column and one output column). Otherwise, you should define one big `MapValues` transformation.

The fact that you're dealing with a 100'000-element lookup table is a bit concerning, because it would probably require some special tweaking/customization at the PMML consumer side (such as the JPMML-Evaluator library) in order to ensure the best performance.

As for conversion from Scikit-Learn representation to PMML representation, then it's a low-level technicality. What Scikit-Learn classes do you use for that in your application code? Can you provide a working example using some toy dataset?

**Comment by jibybabu** [comment]
Hi Vilu,
Thanks for the quick turn around! excited to see your quick response!

So here is my story,
I am building a model to solve a classification problem and needs to convert it into JPMML for deployment.
One of model feature is a string, say name of a person, and my model looks for 10 additional features based numerical(decimal) values based on the name of person. I have kept this map in a CSV file (Name, Col1, Col2, Col 3....Col10) with rows for example("ABC",1.534346,5.1232343,....).
So whenever a observation comes for prediction, the model needs to look for the respective 10 numerical values and use them for scoring the model. 

Right now, i haven't used any skikitlearn transformations to map this, Just populating the column values using python by reading the lookup.csv file. And for deploying, converting into JPMML , it looks like a challenge now. I am ready to change the implementation. 

Any suggestions/advices are much appreciated!

Thanks,
Jiby

**Comment by vruusmann** [comment]
The `MapValues` transformation is designed to mapping many input values to **one** output values. You're interested in achieving exactly the opposite - mapping one input value (eg. entity's identifier) to **many** output values (eg. entity's descriptors).

You could achieve this by defining many `MapValue` transformations - one for each output dimension. This could work if the size of input value set is relatively small and fixed.

Is the "input value set is fixed" requirement met in your use case? In other words, are you going to make predictions only about persons that were known at the time when the model was trained (and converted to PMML). Aren't you interested in making predictions about new persons?

I'd recommend using a layered approach, where the lookup table functionality is moved outside of the PMML document, to a specialized service. The latter could be some REST web service, or an SQL database call. You would be performing this call using a custom Java user-defined function (UDF):
```
<Apply function="com.mycompany.rest.PersonService">
  <FieldRef name="Id"/> <!-- Entity identifier -->
  <Constant>Col1, Col2, Col3, .., Col10</Constant> <!-- Entity attributes to select -->
</Apply>
```

**Comment by jibybabu** [comment]
Hi Vilu,

Thanks a lot for the quick response!

Is the "input value set is fixed" requirement met in your use case?Aren't you interested in making predictions about new persons?
>>> Yes. As like this example, when a new person comes in future , i need to look into the lookup table and get the respective 10 values. I am taking care of the scenarios where the persons name is not in the lookup table. 

I was thinking of a workaround; Let me know your suggestion on that prespective.
how about creating a customized transformation function, somthing like http://zacstewart.com/2014/08/05/pipelines-of-featureunions-of-pipelines.html,  for each 10 <key-value> mappings, put that into an sklearn pipeline and convert the pipeline into dataframeMapper ?

Btw.. Do you have any example of converting sklearn pipeline to dataframeMapper using sklearn2pmml?

Thank,
Jiby

**Comment by vruusmann** [comment]
> how about creating a customized transformation function, somthing like http://zacstewart.com/2014/08/05/pipelines-of-featureunions-of-pipelines.html, for each 10 mappings, put that into an sklearn pipeline and convert the pipeline into dataframeMapper ?

That would be unnecessarily complex solution. It's possible to define a new Scikit-Learner transformer class and map it directly to any number of columns:
```
person_mapper = DataFrameMapper([
  ((Gender, Age, Height), CSVLookupTable("persons.csv", "ID"))
])
```

Class `CSVLookupTable` would be defined in `sklearn2pmml.preprocessing` module. It takes the name of the filesystem file, and the name of the primary key column as arguments.

> Do you have any example of converting sklearn pipeline to dataframeMapper using sklearn2pmml?

The conversion of Scikit-Learn pipelines is discussed in a separate issue:
https://github.com/jpmml/jpmml-sklearn/issues/3

**Comment by jibybabu** [comment]
Thanks Villu, Thats a great approach! Let me try it out that.

Btw, just to make sure, about passing the "ID", the "ID" should the primary column name of the lookUpTable rather than the respective data set column name correct?
And this will make sure that we are taking care of scenarios outside of the data set right?

Thanks,
Jiby

**Comment by vruusmann** [comment]
You can't try it out, because class `sklearn2pmml.preprocessing.CSVLookupTable` is fictional.

 But the solution to your problem could be implemented like this. If you can implement the Python side of this class, then I can do the rest.

**Comment by jibybabu** [comment]
I implemented that in the python. But the problem is do the variables in the mapper and estimator while doing sklearn2python to be synced? I am not using the input variable, for example the name in the model, but the output variables, for example Gender, Age, Height, in the model

**Comment by vruusmann** [comment]
A typical workflow for implementing up a custom transformer:

1. Create Python class. This should load a CSV file, and insert three columns (Gender, Age, Height) to the resulting Python data matrix that the estimator can see and use.
2. Create Java class. This should load the same CSV file, and generate three `MapValues` transformations. You can communicate the name of the future input column by defining an appropriate "helper attribute" in the above Python class.
3. Update the sklearn2pmml package to include both Python and Java classes.

The Java part might be tricky, especially considering that there's not much documentation about it. But you could take a look at earlier commits that implemented different Scikit-Learn transformers. For example: https://github.com/jpmml/jpmml-sklearn/commit/5c4a181cd70d06c8179586cb3be0c881a9ba02e0

I will be unable to work on this till the 1st of November.

**Comment by jibybabu** [comment]
Ok. Thanks a ton Villu! I can take care of the python. 
Please let me know here, i will check constantly, when it is ready!

Also just on a side note, it will be great if you could implement it in a smarter way, as u r always, sothat any of the future crazy transformations can be taken care if somebody implements the corresponding python class

**Comment by le-vision** [comment]
> The MapValues transformation is designed for mapping many input values to one output value. 

Thanks for pointing me towards MapValues for representing a lookup table using PMML, the InlineTable type looks ideal.

In the application I'm working on the PMML needs to represent only that lookup logic, i.e. map from all combinations of 2 input values (e.g. age, height) to some output/target/label value (e.g. isTall).

Having read the above I now think I should create a MapValues type DerivedField, and simply assign the value of that DerivedField to the output of the pmml model.

Is there an existing/simple/recommended way to produce such a pmml? I can always write some code to produce the inlineTable xml representation from a pandas dataframe or something, but perhaps there's an existing or better solution. Although it feels unnecessary, are there any transformations in sklearn that would produce such a MapValues DerivedField?

I opened this SO question just before I found this thread: http://stackoverflow.com/questions/40498703/how-to-generate-a-pmml-that-represents-a-simple-lookup-table-logic-using-python

Thanks!
